File: 2016/09/rs/src/main.rs
use regex::Regex;

fn get_length(data: &str, recursive: bool, marker_regex: &Regex) -> usize {
    if let Some(cap) = marker_regex.captures(data) {
        let data_length = cap
            .name("length")
            .unwrap()
            .as_str()
            .parse::<usize>()
            .unwrap();
        let data = cap.name("data").unwrap().as_str();
        return cap.name("prior").unwrap().as_str().len()
            + cap
                .name("repeats")
                .unwrap()
                .as_str()
                .parse::<usize>()
                .unwrap()
                * (if recursive {
                    get_length(&data[..data_length], recursive, marker_regex)
                } else {
                    data_length
                })
            + get_length(&data[data_length..], recursive, marker_regex);
    } else {
        data.len()
    }
}

fn solve(data: &str) -> (usize, usize) {
    let marker_regex =
        Regex::new(r"(?P<prior>[A-Z]*)\((?P<length>\d+)x(?P<repeats>\d+)\)(?P<data>.*)").unwrap();
    (
        get_length(data, false, &marker_regex),
        get_length(data, true, &marker_regex),
    )
}

fn get_input(file_path: &String) -> String {
    String::from(std::fs::read_to_string(file_path).expect("Error reading input file!"))
}

fn main() {
    let args: Vec<String> = std::env::args().collect();
    if args.len() != 2 {
        panic!("Please, add input file path as parameter");
    }
    let now = std::time::Instant::now();
    let (part1_result, part2_result) = solve(&get_input(&args[1]));
    let end = now.elapsed().as_secs_f32();
    println!("P1: {}", part1_result);
    println!("P1: {}", part2_result);
    println!();
    println!("Time: {:.7}", end);
}

File: 2016/09/c/main.c
#include <stdio.h>
#include <stdlib.h>
#include <sys/time.h>
#include <string.h>

typedef char *Input;
typedef struct
{
    long part1;
    long part2;
} Results;

typedef enum {
    PRIOR = 0,
    LENGTH = 1,
    REPEATS = 2,
    DATA = 3
} CrawlPhase;

long getLength(char *data, int start, int end, int recursive)
{
    int priorLength = 0, length = 0, repeats = 0;
    CrawlPhase phase = PRIOR;
    char c;
    while (start < end)
    {
        c = data[start];
        switch (phase)
        {
            case PRIOR:
                if (c == '(')
                    phase++;
                else
                    priorLength++;
                break;
            case LENGTH:
                if (c == 'x')
                    phase++;
                else
                    length = length * 10 + c - '0';
                break;
            case REPEATS:
                if (c == ')')
                    phase++;
                else
                    repeats = repeats * 10 + c - '0';
                break;
            case DATA:
                return priorLength + repeats * (recursive ? getLength(data, start, start + length, 1) : length) + getLength(data, start + length, end, recursive);
        }
        start++;
    }
    return priorLength;
}

Results solve(Input input)
{
    int end = strlen(input) - 1;
    return (Results){ getLength(input, 0, end, 0), getLength(input, 0, end, 1)};
}

Input getInput(char *filePath)
{
    FILE *file;
    if ((file = fopen(filePath, "r")) == NULL)
    {
        perror("Error reading input file!\n");
        exit(1);
    }
    fseek(file, 0, SEEK_END);
    long length = ftell(file);
    rewind(file);
    char *content = malloc(length);
    fread(content, 1, length, file);
    fclose(file);
    return content;
}

void freeInput(Input input)
{
}

int main(int argc, char **argv)
{
    if (argc != 2)
    {
        perror("Please, add input file path as parameter");
        exit(1);
    }
    struct timeval starts, ends;
    gettimeofday(&starts, NULL);
    Input input = getInput(argv[1]);
    Results results = solve(input);
    gettimeofday(&ends, NULL);
    freeInput(input);
    printf("P1: %ld\n", results.part1);
    printf("P2: %ld\n\n", results.part2);
    printf("Time: %.7f\n", (double)((ends.tv_sec - starts.tv_sec) * 1000000 + ends.tv_usec - starts.tv_usec) / 1000000);
    return 0;
}
